flocker/node/agents/test/test_blockdevice.py

```diff
@@ -0,0 +1,459 @@
+            b'192.0.2.123'
+        )
+        device_path = self.api.get_device_path(attached_volume.blockdevice_id)
+        self.assertTrue(
+            device_path.isBlockDevice(),
+            u"Not a block device. Path: {!r}".format(device_path)
+        )
+
+    def test_get_device_path_device_exists(self):
+        """
+        ``get_device_path`` returns the same ``FilePath`` for the volume device
+        when called multiple times.
+        """
+        new_volume = self.api.create_volume(size=1024 * 50)
+        attached_volume = self.api.attach_volume(
+            new_volume.blockdevice_id,
+            b'192.0.2.123'
+        )
+
+        device_path1 = self.api.get_device_path(attached_volume.blockdevice_id)
+        device_path2 = self.api.get_device_path(attached_volume.blockdevice_id)
+
+        self.assertEqual(device_path1, device_path2)
+
+
+def make_iblockdeviceapi_tests(blockdevice_api_factory):
+    """
+    :returns: A ``TestCase`` with tests that will be performed on the
+       supplied ``IBlockDeviceAPI`` provider.
+    """
+    class Tests(IBlockDeviceAPITestsMixin, SynchronousTestCase):
+        def setUp(self):
+            self.api = blockdevice_api_factory(test_case=self)
+
+    return Tests
+
+
+def losetup_detach(device_file):
+    """
+    Detach the supplied loopback ``device_file``.
+    """
+    check_output(['losetup', '--detach', device_file.path])
+
+
+def losetup_detach_all(root_path):
+    """
+    Detach all loop devices associated with files contained in ``root_path``.
+
+    :param FilePath root_path: A directory in which to search for loop device
+        backing files.
+    :param list backing_files: A ``list`` of all loopback backing files.
+    """
+    for device_file, backing_file in losetup_list():
+        try:
+            backing_file.segmentsFrom(root_path)
+        except ValueError:
+            pass
+        else:
+            losetup_detach(device_file)
+
+
+def loopbackblockdeviceapi_for_test(test_case):
+    """
+    :returns: A ``LoopbackBlockDeviceAPI`` with a temporary root directory
+        created for the supplied ``test_case``.
+    """
+    user_id = os.getuid()
+    if user_id != 0:
+        raise SkipTest(
+            "``LoopbackBlockDeviceAPI`` uses ``losetup``, "
+            "which requires root privileges. "
+            "Required UID: 0, Found UID: {!r}".format(user_id)
+        )
+
+    root_path = test_case.mktemp()
+    test_case.addCleanup(losetup_detach_all, FilePath(root_path))
+    return LoopbackBlockDeviceAPI.from_path(root_path=root_path)
+
+
+class LoopbackBlockDeviceAPITests(
+        make_iblockdeviceapi_tests(
+            blockdevice_api_factory=loopbackblockdeviceapi_for_test
+        )
+):
+    """
+    Interface adherence Tests for ``LoopbackBlockDeviceAPI``.
+    """
+
+
+class LoopbackBlockDeviceAPIImplementationTests(SynchronousTestCase):
+    """
+    Implementation specific tests for ``LoopbackBlockDeviceAPI``.
```

Comment:
The tests were useful during TDD of the implementation, so they may as well stay, right?